src/hotspot/cpu/ppc/upcallLinker_ppc.cpp

```diff
@@ -23,15 +23,337 @@
+  //      | (optional)          |
+  //      | ret_buf             |
+  //      |---------------------| = ret_buf_offset
+  //      |                     |
+  //      | FrameData           |
+  //      |---------------------| = frame_data_offset
+  //      |                     |
+  //      | reg_save_area       |
+  //      |---------------------| = reg_save_are_offset
+  //      |                     |
+  //      | arg_save_area       |
+  //      |---------------------| = arg_save_are_offset
+  //      |                     |
+  //      | res_save_area       |
+  //      |---------------------| = res_save_are_offset
+  //      |                     |
+  // SP-> | out_arg_area        |   needs to be at end for shadow space
+  //
+  //
+
+  //////////////////////////////////////////////////////////////////////////////
+
+  MacroAssembler* _masm = new MacroAssembler(&buffer);
+  address start = __ function_entry(); // called by C
+  __ save_LR_CR(R0);
+  assert((abi._stack_alignment_bytes % 16) == 0, "must be 16 byte aligned");
+  // allocate frame (frame_size is also aligned, so stack is still aligned)
+  __ push_frame(frame_size, tmp);
+
+  // we have to always spill args since we need to do a call to get the thread
+  // (and maybe attach it).
+  arg_spiller.generate_spill(_masm, arg_save_area_offset);
+  // Java methods won't preserve them, so save them here:
+  preserve_callee_saved_registers(_masm, abi, reg_save_area_offset);
+
+  // Java code uses TOC (pointer to code cache).
+  __ load_const_optimized(R29_TOC, MacroAssembler::global_toc(), R0); // reinit
+
+  __ block_comment("{ on_entry");
+  __ load_const_optimized(call_target_address, CAST_FROM_FN_PTR(uint64_t, UpcallLinker::on_entry), R0);
+  __ addi(R3_ARG1, R1_SP, frame_data_offset);
+  __ call_c(call_target_address);
+  __ mr(R16_thread, R3_RET);
+  __ block_comment("} on_entry");
+
+  __ block_comment("{ argument shuffle");
+  arg_spiller.generate_fill(_masm, arg_save_area_offset);
+  if (needs_return_buffer) {
+    assert(ret_buf_offset != -1, "no return buffer allocated");
+    __ addi(as_Register(locs.get(StubLocations::RETURN_BUFFER)), R1_SP, ret_buf_offset);
+  }
+  __ ld(callerSP, _abi0(callers_sp), R1_SP); // preset (used to access caller frame argument slots)
+  arg_shuffle.generate(_masm, as_VMStorage(callerSP), frame::abi_minframe_size, frame::jit_out_preserve_size, locs);
+  __ block_comment("} argument shuffle");
+
+  __ block_comment("{ receiver ");
+  __ load_const_optimized(R3_ARG1, (intptr_t)receiver, R0);
+  __ resolve_jobject(R3_ARG1, tmp, R31, MacroAssembler::PRESERVATION_FRAME_LR_GP_FP_REGS); // kills R31
+  __ block_comment("} receiver ");
+
+  __ load_const_optimized(R19_method, (intptr_t)entry);
+  __ std(R19_method, in_bytes(JavaThread::callee_target_offset()), R16_thread);
+
+  __ ld(call_target_address, in_bytes(Method::from_compiled_offset()), R19_method);
+  __ mtctr(call_target_address);
+  __ bctrl();
```

Comment:
There's no instruction which can use a GP reg as branch target directly. That's why we use CTR.
In this case, using R12 is not required since we call Java and our PPC64 VM code does not rely on it.
If we were calling C, using R12 would be required by ABI v2.